### ISSUE TITLE
No error message received when adding up to 2 letters at the end of phone number string in the sign in page

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#19751]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the "Staging New Expensify" website.
2. Input the phone number and enter any alphabets at the end.
3. Observe there is no error message and user taken to magic code page
4. Now add the 3rd letter at the end

Expected result:
Either error message should be displayed when adding letters at the end of phone number string in step 3
Letters should be truncated when navigated to magic code page

Actual result:
No error message until adding 2 letters, taken to magic code page and number entered is trimmed
But if 3rd letter added error message appears

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.19-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/e0fef2ce-c869-4925-93e2-1371d3acaf7f

https://github.com/Expensify/App/assets/43996225/fea72946-4e16-46e2-8bd8-e662ea1e4cf9

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Oleksii-Bashkyrov
**Slack conversation:** 
https://expensify.slack.com/archives/C049HHMV9SM/p1683707368642449
[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by qasimkhan23** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
No error message until adding 2 letters, taken to magic code page and number entered is trimmed
But if 3rd letter added error message appears

### What is the root cause of that problem?
The issue is from the function parsedPhoneNumber which we are using from a lib "awesome-phonenumber", this function return possible : true if 1 or 2 characters are added at the end and we have a condition against this property so the error doesn't pop.
https://github.com/Expensify/App/blob/f55c36ca9be7f347713a89521826449d90d07d06/src/pages/signin/LoginForm.js#L149-L158


<img width="1435" alt="Screenshot 2023-05-29 at 11 35 15 PM" src="https://github.com/Expensify/App/assets/46372909/fcd28730-8ff5-4534-b6da-1d5e601bb38b">

### What changes do you think we should make in order to solve the problem?

We can solve this by replacing the extra added characters if there are any with the empty string if the property possible is true otherwise the the state would be update with the original number which was added first.

```
if (!Str.isValidEmail(login) && !parsedPhoneNumber.possible) {
            if (ValidationUtils.isNumericWithSpecialChars(login)) {
                this.setState({formError: 'common.error.phoneNumber'});
            } else {
                this.setState({formError: 'loginForm.error.invalidFormatEmailLogin'});
            }
            return;
        }else{
            this.setState({
                login:parsedPhoneNumber.possible ? login.replace(/\D+/g, '') : login
            });
        }
```
the extra characters will be removed as soon as the user clicks the continue button.
After applying the solution

https://github.com/Expensify/App/assets/46372909/22f1f515-7965-42aa-81e0-ed7162e828e0

### What alternative solutions did you explore? (Optional)
We can open an issue in github repo of the library we are using "awesome-phonenumber" and wait to get this fixed.

**Comment by jliexpensify** [comment]
I can repro:

https://github.com/Expensify/App/assets/30815269/cbe6d5ab-3d2c-4d00-ba42-cb722042f841

I think for consistency's sake, we should either:

1. Surface the error if ant number of letters are added after a number OR
2. Automatically remove any letters after a number (since this isn't required)

**Comment by jliexpensify** [comment]
Asking internally here - https://expensify.slack.com/archives/C01SKUP7QR0/p1685405176413529

**Comment by jliexpensify** [comment]
Got some consensus on the behaviour we want:

1. We should not change the user’s input
2. Always throw an error if what they input is invalid

So this means, if they enter even 1 letter after a number, we need to throw the `Invalid` error.

**Comment by sethraj14** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

No error message is shown when adding up to 2 letters at the end of the phone number input on sign-in page.

### What is the root cause of that problem?

The library we are using to parse phone number and get if is valid or not is parsing the last two characters internally and giving us validation of the phone number. 

Example response - 
```javascript
{
    "number": {
        "input": "+14098797654ab",
        "international": "+1 409-879-7654",
        "national": "(409) 879-7654",
        "e164": "+14098797654",
        "rfc3966": "tel:+1-409-879-7654",
        "significant": "4098797654"
    },
    "regionCode": "US",
    "valid": true,
    "possible": true,
    "canBeInternationallyDialled": true,
    "possibility": "is-possible",
    "type": "fixed-line-or-mobile",
    "typeIsMobile": true,
    "typeIsFixedLine": true,
    "countryCode": 1
}
``` 



### What changes do you think we should make in order to solve the problem?
Keeping @jliexpensify points in mind that we do not need to change the user's input and show an error on any character addition after phone no, we can put regex phone check on phone input along with the library valid check like this - 

```javascript

        const phoneValidationRegex = /^[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}$/;

        if (!Str.isValidEmail(login) && (!parsedPhoneNumber.possible || !phoneValidationRegex.test(phoneLogin))) {

``` 

This will ensure that all the previous check works according to the library along with this new check on the phone input.

### What alternative solutions did you explore? (Optional)

We can look for an overall solution for the library `awesome-phonenumber` parsePhoneNumber method.

**Comment by rnsshrmb2010** [comment]
Replace code 

const phoneLogin = LoginUtils.appendCountryCode(LoginUtils.getPhoneNumberWithoutSpecialChars(login));
        const parsedPhoneNumber = parsePhoneNumber(phoneLogin);

        if (!Str.isValidEmail(login) && !parsedPhoneNumber.possible) {
            if (ValidationUtils.isNumericWithSpecialChars(login)) {
                this.setState({formError: 'common.error.phoneNumber'});
            } else {
                this.setState({formError: 'loginForm.error.invalidFormatEmailLogin'});
            }
            return;
        }


With 

const phoneWithSpChar = LoginUtils.getPhoneNumberWithoutSpecialChars(login)

        const phoneLogin = LoginUtils.appendCountryCode(phoneWithSpChar);
        const parsedPhoneNumber = parsePhoneNumber(phoneLogin);

        if (!Str.isValidEmail(login) && (!parsedPhoneNumber.possible || Number.isNaN(Number(phoneWithSpChar)))) {
            if (!ValidationUtils.isNumericWithSpecialChars(login)) {
                this.setState({formError: 'common.error.phoneNumber'});
            } else {
                this.setState({formError: 'loginForm.error.invalidFormatEmailLogin'});
            }
            return;
        }


It will work perfectly without changing your existing login

**Comment by rnsshrmb2010** [comment]
Contribute Details
Expensify account email: rnsshrmb2010@gmail.com
Upwork Profile Link: [<REPLACE LINK HERE>](https://www.upwork.com/freelancers/~011287a1405a39b39d)

**Comment by Santhosh-Sellavel** [comment]
@hoangzinh Please check the recording in the issue [thread](https://github.com/Expensify/App/issues/19751#issue-1731066175). I tried the same.

**Comment by Santhosh-Sellavel** [comment]
Alright adding a plus help easily reproduce

**Comment by Santhosh-Sellavel** [comment]
May be this is intentional I guess https://github.com/google/libphonenumber/blob/master/FAQ.md#non_digits

**Comment by qasimkhan23** [comment]
In that case we should remove the extra added character which could be a typo and update the state before going onto the next screen

**Comment by qasimkhan23** [comment]
But if you click on the go back button from the magic code screen, you would not see the typo, which otherwise would be there and can be misleading

**Comment by stitesExpensify** [comment]
So if according to awesome-phonenumber this is intentional, then I guess it's intentional? Maybe there are phone numbers like that? I really don't know

**Comment by sethraj14** [comment]
Going by the consensus as @jliexpensify mentioned that we should not automatically change the user's input without notifying the user about the error. 

@stitesExpensify After checking the reason in the awesome-phonenumber library [here](https://github.com/google/libphonenumber/blob/master/FAQ.md#why-does-the-library-treat-some-non-digit-characters-as-digits, it mentions that library doesn't aim to verify that string is only a phone number. It extracts the phone number from the given string. 

Hence we should put a check from our side to the given input and show an error to the user in case of an error.  

cc: @Julesssss @Santhosh-Sellavel

**Comment by rnsshrmb2010** [comment]
Hi Team,

I analyzed your issue and why it is not triggering.

*parsePhoneNumber(phoneLogin)* not validating if you add two alphabet
characters  at the end and it is provided as status "is-possible". If you
once more 3 alphabets at the end then it is providing invalid as status.

To fix this issue you have added one more condition and need to check if
the given input is number or not after removing the special character. Then
only it will trigger a message for the invalid phone number.

And more issue in logic is that,
*ValidationUtils.isNumericWithSpecialChars(login)
*line returns *TRUE *if the given input number is valid. Hear your ae
checking if given input number is valid then show error message.

Because of the above issues in your code, the message does not trigger an
error message.

Please follow my steps to fix this issue



Replace your code from line number 148 to 158


const phoneLogin =
LoginUtils.appendCountryCode(LoginUtils.getPhoneNumberWithoutSpecialChars(login));
        const parsedPhoneNumber = parsePhoneNumber(phoneLogin);

        if (!Str.isValidEmail(login) && !parsedPhoneNumber.possible) {
            if (ValidationUtils.isNumericWithSpecialChars(login)) {
                this.setState({formError: 'common.error.phoneNumber'});
            } else {
                this.setState({formError:
'loginForm.error.invalidFormatEmailLogin'});
            }
            return;
        }



With below code

      const phoneWithSpChar =
LoginUtils.getPhoneNumberWithoutSpecialChars(login)

        const phoneLogin = LoginUtils.appendCountryCode(phoneWithSpChar);
        const parsedPhoneNumber = parsePhoneNumber(phoneLogin);

        if (!Str.isValidEmail(login) && (!parsedPhoneNumber.possible ||
Number.isNaN(Number(phoneWithSpChar)))) {
            if (!ValidationUtils.isNumericWithSpecialChars(login)) {
                this.setState({formError: 'common.error.phoneNumber'});
            } else {
                this.setState({formError:
'loginForm.error.invalidFormatEmailLogin'});
            }
            return;
        }




Regards,
Hanamantraya Rajashekhar
9582017725
Email : ***@***.***

**Comment by Santhosh-Sellavel** [comment]
@stitesExpensify 

> So if according to awesome-phonenumber this is intentional, then I guess it's intentional? Maybe there are phone numbers like that? I really don't know

Have a look at this 
> 
> ### [Why does the library treat some non-digit characters as digits?](https://github.com/google/libphonenumber/blob/master/FAQ.md#non_digits)
> When parsing, the library does its best to extract a phone number out of the given input string. It looks for the phone number in the provided text; it doesn't aim to verify whether the string is only a phone number.
> 
> If the input looks like a vanity number to the library, parse() assumes this is intentional and converts alpha characters to digits. Please read the documentation for PhoneNumber parse(String, String) in [PhoneNumberUtil](http://github.com/google/libphonenumber/blob/master/java/libphonenumber/src/com/google/i18n/phonenumbers/PhoneNumberUtil.java) for details. Also see Iterable<PhoneNumberMatch> findNumbers(CharSequence, String).
> 
> Some examples:
> 
> +1 412 535 abcd is parsed the same as +1 412 535 2223.
> 
> If someone mistypes and adds an extra alpha character in the middle, then the library assumes this was a mistake and fixes it. E.g. the extra c in +1 412 535 c0000 is ignored, and this is parsed the same as +1 412 535 0000.
> 
> If someone mistypes and replaces a digit in the middle with an alpha character, and the remaining characters do not make up a valid number, this alpha character is not converted and the resulting number is invalid, e.g. with +1 412 535 c000.
> 
> 


After seeing this the library does it intentionally to ignore typos. Adding any additional validation checks makes the use of the library pointless here.

**Comment by Santhosh-Sellavel** [comment]
@rnsshrmb2010 Please start by reading our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md). Understand our process,  and submit proposal following our guidelines & proposal template.

**Comment by rnsshrmb2010** [comment]
By using this library this issue can't fix. 

@Santhosh-Sellavel 
Sure I will read your guide line. I want to know one thing is that weather we have to use that library hear or we have find out some other way to fix issue?

**Comment by rnsshrmb2010** [comment]
By using this library this issue can't fix.

@Santhosh-Sellavel
Sure I will read your guide line. I want to know one thing is that weather
we have to use that library hear or we have find out some other way to fix
issue?

Shall commit my code and I will create PR to merge code?

>

**Comment by Santhosh-Sellavel** [comment]
@rnsshrmb2010 Don't rush again we are discussing to decide whether we want to handle this issue.

Please start reading our [contributing guidelines] (https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) first, you will understand the process which clears all your queries.

**Comment by rnsshrmb2010** [comment]
Sure I am processing step by step

**Comment by rnsshrmb2010** [comment]
My account created on slack and I want to know what the next process is?
please guide start work on this issue

**Comment by Santhosh-Sellavel** [comment]
> @stitesExpensify
> 
> > So if according to awesome-phonenumber this is intentional, then I guess it's intentional? Maybe there are phone numbers like that? I really don't know
> 
> Have a look at this
> 
> > ### [Why does the library treat some non-digit characters as digits?](https://github.com/google/libphonenumber/blob/master/FAQ.md#non_digits)
> > When parsing, the library does its best to extract a phone number out of the given input string. It looks for the phone number in the provided text; it doesn't aim to verify whether the string is only a phone number.
> > If the input looks like a vanity number to the library, parse() assumes this is intentional and converts alpha characters to digits. Please read the documentation for PhoneNumber parse(String, String) in [PhoneNumberUtil](https://github.com/google/libphonenumber/blob/master/java/libphonenumber/src/com/google/i18n/phonenumbers/PhoneNumberUtil.java) for details. Also see Iterable findNumbers(CharSequence, String).
> > Some examples:
> > +1 412 535 abcd is parsed the same as +1 412 535 2223.
> > If someone mistypes and adds an extra alpha character in the middle, then the library assumes this was a mistake and fixes it. E.g. the extra c in +1 412 535 c0000 is ignored, and this is parsed the same as +1 412 535 0000.
> > If someone mistypes and replaces a digit in the middle with an alpha character, and the remaining characters do not make up a valid number, this alpha character is not converted and the resulting number is invalid, e.g. with +1 412 535 c000.
> 
> After seeing this the library does it intentionally to ignore typos. Adding any additional validation checks makes the use of the library pointless here.

@stitesExpensify @Julesssss thougths?

**Comment by stitesExpensify** [comment]
> After seeing this the library does it intentionally to ignore typos. Adding any additional validation checks makes the use of the library pointless here.

I disagree. I think that we should still validate that there are no letters. Doing additional validation does not remove the usefulness of the library validation.

**Comment by Julesssss** [comment]
Yep, lets still resolve on our side.

**Comment by rnsshrmb2010** [comment]
[Julesssss](https://github.com/Julesssss) Can I resolve by adding some extra condition on it?

**Comment by sethraj14** [comment]
@Julesssss Can you review my [proposal](https://github.com/Expensify/App/issues/19751#issuecomment-1569258531), I have followed the given guidelines and conditions.

**Comment by qasimkhan23** [comment]
> ## Proposal
> 
> ### Please re-state the problem that we are trying to solve in this issue.
> No error message until adding 2 letters, taken to magic code page and number entered is trimmed
> But if 3rd letter added error message appears
> 
> ### What is the root cause of that problem?
> The issue is from the function parsedPhoneNumber which we are using from a lib "awesome-phonenumber", this function return possible : true if 1 or 2 characters are added at the end and we have a condition against this property so the error doesn't pop.
> https://github.com/Expensify/App/blob/f55c36ca9be7f347713a89521826449d90d07d06/src/pages/signin/LoginForm.js#L149-L158
> 
> 
> <img width="1435" alt="Screenshot 2023-05-29 at 11 35 15 PM" src="https://github.com/Expensify/App/assets/46372909/fcd28730-8ff5-4534-b6da-1d5e601bb38b">
> 
> ### What changes do you think we should make in order to solve the problem?
> 
> We can solve this by replacing the extra added characters if there are any with the empty string if the property possible is true otherwise the the state would be update with the original number which was added first.
> 
> ```
> if (!Str.isValidEmail(login) && !parsedPhoneNumber.possible) {
>             if (ValidationUtils.isNumericWithSpecialChars(login)) {
>                 this.setState({formError: 'common.error.phoneNumber'});
>             } else {
>                 this.setState({formError: 'loginForm.error.invalidFormatEmailLogin'});
>             }
>             return;
>         }else{
>             this.setState({
>                 login:parsedPhoneNumber.possible ? login.replace(/\D+/g, '') : login
>             });
>         }
> ```
> the extra characters will be removed as soon as the user clicks the continue button.
> After applying the solution
> 
> https://github.com/Expensify/App/assets/46372909/22f1f515-7965-42aa-81e0-ed7162e828e0
> 
> Alternatively we can add a regex for validation if 1 or more characters are added at the end
> like this
> 
> ```
>    const customPhoneValidationRegex = /^[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}$/;
> 
>         if (!Str.isValidEmail(login) && !parsedPhoneNumber.possible || !customPhoneValidationRegex.test(phoneLogin)) {
>             if (ValidationUtils.isNumericWithSpecialChars(login)) {
>                 this.setState({formError: 'common.error.phoneNumber'});
>             } else {
>                 this.setState({formError: 'loginForm.error.invalidFormatEmailLogin'});
>             }
>             return;
>         }
> ``` 
> After applying this solution 
> <img width="1440" alt="Screenshot 2023-05-31 at 10 34 00 AM" src="https://github.com/Expensify/App/assets/46372909/7f4ced93-4186-4fb4-8a71-d667de765dde">
> 
> 
> ### What alternative solutions did you explore? (Optional)
> We can open an issue in github repo of the library we are using "awesome-phonenumber" and wait to get this fixed.

@Santhosh-Sellavel the regex here validates if the user has added the extra character and gives the validation error

**Comment by Santhosh-Sellavel** [comment]
@qasimkhan23 You have just modified your solution post-submission of the @sethraj14's proposal, your initial solution never  had regex.

**Comment by qasimkhan23** [comment]
@Santhosh-Sellavel I didn't know we can't modify after submission, I'll keep that in mind next time

**Comment by Santhosh-Sellavel** [comment]
We can modify the solution, but should not be a duplicate of other proposals. Also when you modify a solution should post a comment that the proposal updated.

**Comment by sethraj14** [comment]
@Santhosh-Sellavel The regex validates all the phone number inputs. It catches our issue and gives the required error to the user. As discussed [here](https://github.com/Expensify/App/issues/19751#issuecomment-1569220040), we do not automatically want to modify the user input, this regex works with our library check.

**Comment by Santhosh-Sellavel** [comment]
@sethraj14
> The regex validates all the phone number inputs.

Can you explain/elaborate on how this regex works, is this any standard regex?

**Comment by sethraj14** [comment]
> @sethraj14
> 
> > The regex validates all the phone number inputs.
> 
> Can you explain/elaborate on how this regex works, is this any standard regex?


Yes, the regex currently matches with all these possibilities of the phone number. 

![Screenshot 2023-06-06 at 11 15 36 PM](https://github.com/Expensify/App/assets/14844259/7e6e53dc-ef96-40ae-800a-c00d2d1d2384)

There is more extensive regex for phone number matching but as we are also using awesome-phonenumber library, this regex will be verifying the contents of the phone number.

We can surely modify the regex in the PR.

Alternate regex 
`^(\+\d{1,2}\s?)?1?\-?\.?\s?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4}$\`

**Comment by sethraj14** [comment]
@Santhosh-Sellavel Detailed explanation of both regex - 
`/^[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}$/`

![Screenshot 2023-06-06 at 11 26 07 PM](https://github.com/Expensify/App/assets/14844259/d4492da7-5841-49ea-bd90-c64c2da62228)

`^(\+\d{1,2}\s?)?1?\-?\.?\s?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4}$\`

![Screenshot 2023-06-06 at 11 25 31 PM](https://github.com/Expensify/App/assets/14844259/5d4d1ec7-79b4-4d63-a6b5-dfdd176313c7)

**Comment by Santhosh-Sellavel** [comment]
@Julesssss Are we good with adding yet another regex for  Phone validation?

**Comment by Julesssss** [comment]
> Are we good with adding yet another regex for Phone validation?

I'm struggling to look for our existing phone regex (other than the lib), but I think it's okay in this case as long as we document why we needed this new regex.

**Comment by Santhosh-Sellavel** [comment]
@Prince-Mendiratta what was the need for us to include the library? To handle what special cases?

**Comment by Julesssss** [comment]
> I think we can use this existing simple regex to ensure it only includes digits then use exiting lib to validate possibility of it.

@hoangzinh  I like the sound of this, though my only remaining question is whether the existing library would still be necessary.

**Comment by Santhosh-Sellavel** [comment]
> > I think we can use this existing simple regex to ensure it only includes digits then use exiting lib to validate possibility of it.
> 
> @hoangzinh I like the sound of this, though my only remaining question is whether the existing library would still be necessary.

@Julesssss  If we use the existing phone regex we need to remove the library as it adds no values. 
Ex: Using the library we expect +1.800.555.1234 as a valid phone number, but as per our existing regex it's not valid.

@stitesExpensify @Prince-Mendiratta Can you help us this move forward? 

- What was the need for us to include the library? 
- To handle what special cases?
- What characters are allowed as valid for phone number input?

cc: @Julesssss

**Comment by Julesssss** [comment]
I'm leaning towards replacing the library, but it would be good to clarify the questions above first so that we can be sure we're not losing any benefits the lib provides. I also asked in Slack [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1686555416737809)

**Comment by Julesssss** [comment]
Thank you. Maybe the simplest solution is to just close the issue and do nothing here then.

**Comment by s77rt** [comment]
Please check https://github.com/Expensify/App/issues/16760#issuecomment-1564829986 (second paragraph)

TL; DR: A library feature/limitation. Probably nothing to do unless we can:
1. Disable the feature (need to read the above comment)
2. Replace the library

**Comment by Julesssss** [comment]
Okay, lets close this.
